Always display Find Online Subtitle OSD messages

low-batt opened the pull request:
This commit will:
- Add an isAlwaysShown property to OSDMessage that is true for OSD messages emitted by the Find Online Subtitles feature
- Change PlayerCore.sendOSD to send messages whose isAlwaysShown property is true even when the enableOSD setting is false
- Change MainWindowController.displayOSD to display messages whose isAlwaysShown property is true even when the IINA OSD is disabled due to the user enabling mpv's OSD

These changes correct a problem where IINA's Find Online Subtitles feature appears to do nothing when the OSD is disabled. This feature uses the OSD to interact with the user. Failing to display the messages breaks the feature.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [ ] This implements/fixes issue #.

---

**Description:**

low-batt committed:
Always display Find Online Subtitle OSD messages

This commit will:
- Add an isAlwaysShown property to OSDMessage that is true for OSD
  messages emitted by the Find Online Subtitles feature
- Change PlayerCore.sendOSD to send messages whose isAlwaysShown
  property is true even when the enableOSD setting is false
- Change MainWindowController.displayOSD to display messages whose
  isAlwaysShown property is true even when the IINA OSD is disabled due
  to the user enabling mpv's OSD

These changes correct a problem where IINA's Find Online Subtitles
feature appears to do nothing when the OSD is disabled. This feature
uses the OSD to interact with the user. Failing to display the messages
breaks the feature.

low-batt commented:
This does not address `customWithDetail` which is used by plugins. I'm guessing there may be a similar problem with messages emitted by plugins.

Thoughts?

lhc70000 commented:
`isAlwaysShown` looks confusing to me. Initially I thought it would make the OSD persistently displayed on the screen and not hiding anymore. Maybe it can be something more straightforward, like `alwaysEnabled` or `ignoreDisableOSD`?

low-batt committed:
Always display Find Online Subtitle OSD messages

This commit will:
- Add an alwaysEnabled property to OSDMessage that is true for OSD
  messages emitted by the Find Online Subtitles feature
- Change PlayerCore.sendOSD to send messages whose alwaysEnabled
  property is true even when the enableOSD setting is false
- Change MainWindowController.displayOSD to display messages whose
  alwaysEnabled property is true even when the IINA OSD is disabled
  due to the user enabling mpv's OSD

These changes correct a problem where IINA's Find Online Subtitles
feature appears to do nothing when the OSD is disabled. This feature
uses the OSD to interact with the user. Failing to display the messages
breaks the feature.

low-batt commented:
Makes sense. I was too close to the code to notice that it could be interpreted that way. I renamed it to `alwaysEnabled`.

We still need to figure out how to properly solve the similar issue for plugins.

svobs commented:
> Thoughts?

In PR #4991 you wrote this:

> A solution that exposes all (excluding the mandatory ones) OSD messages really should be data driven, so adding a new enum to OSDMessage automatically appears in the UI. Very doable, but a lot more work. This also means the enum can't have "generic" messages that actually represent multiple different kinds of messages like the current "custom" entries do.

This sounds like you have some idea for what you would do for a more comprehensive fix? To throw in my suggestion, I think it would be less code to maintain in the long-term if, instead of adding booleans to `OSDMessage`, just add a `canShow(msg)` func prior to `displayOSD` which will discard the message if it can't be shown. That way, more complex logic can be used, e.g., a switch which checks for all the messages to deny when `PK.enableOSD` is false, in addition to checks for all the individually disabled messages, and the logic will all be in one place. Any added enums could simply default to `true` unless code is added there. Enums for plugins could attach a property containing their own enablement state or specify a block of code to execute to determine it.

Of course if this is just a stopgap fix, then hard to argue with the winding path of progress :)

lhc70000 commented on iina/PlayerCore.swift:
```diff
@@ -2158,7 +2158,7 @@ class PlayerCore: NSObject {
   func sendOSD(_ osd: OSDMessage, autoHide: Bool = true, forcedTimeout: Float? = nil, accessoryView: NSView? = nil, context: Any? = nil, external: Bool = false) {
     // querying `mainWindow.isWindowLoaded` will initialize mainWindow unexpectly
-    guard mainWindow.loaded && Preference.bool(for: .enableOSD) else { return }
+    guard mainWindow.loaded, Preference.bool(for: .enableOSD) || osd.alwaysEnabled else { return }
```
Why using `,` here instead of `&&`?

low-batt replied:
I tried to find what Swift considers the best practice in this case and only found people debating that. The advice in the Medium post [Chaining logical expressions with commas in Swift](https://medium.com/@austinlouden/chaining-logical-expressions-with-commas-in-swift-c3755ee784e9) about  expressions using commas being easier to read than complex expressions reflects what I was thinking. Using `&&` would require adding `()` around the `||` expression. I was also thinking that the comma was more like swift has you have to use it when using `let`.

I'm good with whatever style the group wants to use. Let me know if you want me to switch this to `&&`.

lhc70000 replied:
I was just asking and also found it a little bit confusing. I think it's alright to use `,` here to avoid the parentheses.

low-batt commented:
This was included in IINA 1.4.0.
